diff --git a/asmcomp/closure.ml b/asmcomp/closure.ml
index 1bdc4392e..69f8bd27d 100644
--- a/asmcomp/closure.ml
+++ b/asmcomp/closure.ml
@@ -474,9 +474,9 @@ let simplif_prim_pure fpc p (args, approxs) dbg =
         | Word_size -> make_const_int (8*Arch.size_int)
         | Int_size -> make_const_int (8*Arch.size_int - 1)
         | Max_wosize -> make_const_int ((1 lsl ((8*Arch.size_int) - 10)) - 1 )
-        | Ostype_unix -> make_const_bool (Sys.os_type = "Unix")
-        | Ostype_win32 -> make_const_bool (Sys.os_type = "Win32")
-        | Ostype_cygwin -> make_const_bool (Sys.os_type = "Cygwin")
+        | Ostype_unix -> make_const_bool true
+        | Ostype_win32 -> make_const_bool false
+        | Ostype_cygwin -> make_const_bool false
         | Backend_type ->
             make_const_ptr 0 (* tag 0 is the same as Native here *)
       end
diff --git a/asmcomp/cmmgen.ml b/asmcomp/cmmgen.ml
index 2120d3985..99fc968f7 100644
--- a/asmcomp/cmmgen.ml
+++ b/asmcomp/cmmgen.ml
@@ -1994,9 +1994,9 @@ and transl_prim_1 env p arg dbg =
         | Int_size -> tag_int (Cconst_int ((8*Arch.size_int) - 1)) dbg
         | Max_wosize ->
             tag_int (Cconst_int ((1 lsl ((8*Arch.size_int) - 10)) - 1 )) dbg
-        | Ostype_unix -> const_of_bool (Sys.os_type = "Unix")
-        | Ostype_win32 -> const_of_bool (Sys.os_type = "Win32")
-        | Ostype_cygwin -> const_of_bool (Sys.os_type = "Cygwin")
+        | Ostype_unix -> const_of_bool true 
+        | Ostype_win32 -> const_of_bool false
+        | Ostype_cygwin -> const_of_bool false
         | Backend_type ->
             tag_int (Cconst_int 0) dbg (* tag 0 is the same as Native here *)
       end
